### DISCUSSION TITLE
Call for user feedback on AutoSampler

### DISCUSSION THREAD

**Discussion by y0z** [optuna/optuna#5747] [General]
[AutoSampler](https://hub.optuna.org/samplers/auto_sampler/) is a new feature that automatically selects a sampler in Optuna.

In this thread, we seek user feedback, requests, and comments to continuously improve AutoSampler.
Please feel free to post your thoughts.

<a href="https://hub.optuna.org/samplers/auto_sampler/"><img width="1341" alt="image" src="https://github.com/user-attachments/assets/92daf54b-a576-4082-8c77-6456f4481b0d"></a>

**Comment by tetsuyasu** [comment]
Can AutoSampler be used with Pruner?

**Reply by y0z** [reply]
Thank you for your question.

The answer is Yes. You can use pruners with AutoSampler.
Note that, as same as the built-in Optuna samplers, AutoSampler supports pruning only for single-objective optimization.

**Comment by db0** [comment]
Does autosampler also use SotA ones like HEBO?

**Reply by y0z** [reply]
Thank you for your question.

Currently, AutoSampler only selects Optuna's built-in Sampler internally.

However, if you want to use HEBO in Optuna, you can use `HEBOSampler`, which is published on OptunaHub.

- [HEBO (Heteroscedastic and Evolutionary Bayesian Optimisation) | OptunaHub](https://hub.optuna.org/samplers/hebo/)
- [hebo/example.py](https://github.com/optuna/optunahub-registry/blob/main/package/samplers/hebo/example.py)

```sh
pip install -r https://hub.optuna.org/samplers/hebo/requirements.txt
```

```python
import optuna
import optunahub


module = optunahub.load_module("samplers/hebo")
HEBOSampler = module.HEBOSampler


def objective(trial: optuna.trial.Trial) -> float:
    x = trial.suggest_float("x", -10, 10)
    y = trial.suggest_int("y", -10, 10)
    return x**2 + y**2


if __name__ == "__main__":
    sampler = HEBOSampler(
        {
            "x": optuna.distributions.FloatDistribution(-10, 10),
            "y": optuna.distributions.IntDistribution(-10, 10),
        }
    )
    study = optuna.create_study(sampler=sampler)
    study.optimize(objective, n_trials=100)
    print(study.best_trial.params)

    fig = optuna.visualization.plot_optimization_history(study)
    fig.write_image("hebo_optimization_history.png")
```

<a href="https://hub.optuna.org/samplers/hebo/"><img width="1348" alt="image" src="https://github.com/user-attachments/assets/f5743b02-ccff-444a-9e86-619fc9a396ce"></a>

**Reply by db0** [reply]
Thanks. I read how I can use HEBO already. It would just be nice to also have an option to tell the autosampler "User sampler1 and sampler2 as well i your rotation"

**Comment by rodrigoscoelho** [comment]
In the AutoSampler examples, I noticed that the number of `n_trials` in `study.optimize` is used to determine which sampler to select. However, how does this functionality apply to studies that use the `ask()` and `tell()` approach? Since the expected number of trials isn't explicitly provided to the optimizer in this method, how would AutoSampler handle it? Thank you.

**Reply by not522** [reply]
AutoSampler switches the sampler during the study using the current number of completed trials instead of `n_trials`. So, it does not need to know the total number of trials. If you want to know which sampler is used, try `optuna.logging.set_verbosity(optuna.logging.DEBUG)`.

```python
import optuna
import optunahub

def objective(trial):
  x = trial.suggest_float("x", -5, 5)
  y = trial.suggest_float("y", -5, 5)
  return x**2 + y**2

optuna.logging.set_verbosity(optuna.logging.DEBUG)
module = optunahub.load_module(package="samplers/auto_sampler")
study = optuna.create_study(sampler=module.AutoSampler())
for _ in range(300):
    trial = study.ask()
    study.tell(trial, objective(trial))
```

**Reply by rodrigoscoelho** [reply]
Thank you very much, not522.

I followed your advice and enabled logging on my autosampler, but for some reason, it consistently defaults to using TPE. The first iteration (iteration 0) is processed with the RandomSampler (enqueued params), but from iteration 1 through evaluation 100, the TPESampler is used.

Here’s an example from the log:
```
Trial 0 popped from the trial queue.
[D 2024-11-22 16:41:07,769] Sample trial#0 with RandomSampler.
Sample trial#1 with TPESampler.
Sample trial#17 with TPESampler.
Sample trial#187 with TPESampler.
```

It also logs trials like `Sample trial#187` because many trials were marked as FAILED due to external constraints failing the sampled parameters. In reality, only 100 trials have been completed successfully.

Do you know why the GP sampler isn’t being used at the start? This is a single-objective optimization problem.
Using Ask() and Tell(). 

Thank you again for your help!

**Comment by rodrigoscoelho** [comment]
I found out that when the constraint_func is enabled, it always uses TPE as the sampler. I believe this information should be included in the instructions.

**Reply by not522** [reply]
Thank you for your feedback! This behavior is intended, as noted in [the blog post](https://medium.com/optuna/autosampler-automatic-selection-of-optimization-algorithms-in-optuna-1443875fd8f9). It may be updated in the future (for example, `GPSampler` will be one of the choices for constrained optimizations).

> I believe this information should be included in the instructions.

Thanks. I think the document has room for improvement.

**Reply by rodrigoscoelho** [reply]
Maybe an explanation like this:

Initial State:
The sampler starts as a RandomSampler by default.

Multi-Objective Optimization:
If the study has more than one objective (len(study.directions) > 1):
NSGAIISampler is chosen for multi-objective optimization.

Single-Objective Optimization:
If the study has only one objective, the following conditions are considered:
a. If constraints_func is provided, or if there are categorical parameters, or if there are conditional parameters:
TPESampler is chosen with multivariate=True and constant_liar=True.

b. If the search space is purely numerical (no categorical or conditional parameters):
If the number of completed trials is less than 250 (self._N_COMPLETE_TRIALS_FOR_CMAES):
GPSampler is chosen.

If the number of completed trials is 250 or more:
If the search space has more than one parameter:
CmaEsSampler is chosen, warm-started with the first 250 completed trials.
If the search space has only one parameter:
TPESampler is chosen with multivariate=True and constant_liar=True.

Here's a summary of the conditions and corresponding samplers:

Multi-objective: NSGAIISampler
Single-objective with constraints or categorical/conditional parameters: TPESampler
Single-objective, numerical-only search space:

Less than 250 completed trials: GPSampler
250 or more completed trials:

Multiple parameters: CmaEsSampler
Single parameter: TPESampler

The AutoSampler will switch between these samplers as the optimization progresses and the conditions change. It starts with RandomSampler and then adapts based on the study's characteristics and the number of completed trials.

**Comment by lkaino** [comment]
I'm now testing it, sometimes it uses GPSampler in the beginning, sometimes starts with TPE. I don't have any categorical parameters (converted one from bool to int 0..1). It would be nice if the debug log printed the reason for choosing TPE over GP?

**Comment by rodrigoscoelho** [comment]
It actually has to start with GP based on your info. Maybe you are using
constraints func.


 # AutoSampler in Optuna: Summary of Sampler Selection Criteria

The `AutoSampler` in Optuna is designed to automatically select the most appropriate sampler based on the problem's characteristics and the search space. It intelligently switches between different samplers as more information becomes available throughout the optimization process. Below is a summary of when each sampler is selected.

## Overview of Sampler Selection

The `AutoSampler` chooses a sampler based on:
1. **Type of Optimization**: Single-objective vs. Multi-objective.
2. **Characteristics of the Search Space**: Numerical, categorical parameters, conditional parameters, or number of completed trials.

The samplers used are:
- `RandomSampler`
- `TPESampler`
- `GPSampler`
- `CmaEsSampler`
- `NSGAIISampler`

### 1. Multi-Objective Problems
- **Sampler Used**: `NSGAIISampler`
- **Reason**: It is well-suited for multi-objective optimization, leveraging NSGA-II (Non-dominated Sorting Genetic Algorithm II).

### 2. Single-Objective Problems
The choice depends on the characteristics of the search space and the number of completed trials.

#### Initial Sampler
- **`RandomSampler`** is used initially to collect diverse points from the search space before enough information is available.

#### `TPESampler`
- Used when:
  - The search space includes **categorical parameters**.
  - There are **conditional parameters**.
  - **Constraints** are provided via `constraints_func`.
  - Alternatively, used when there are **fewer than 250 trials** completed for numerical-only search spaces.

#### `GPSampler`
- Used when:
  - The search space is **numerical-only**, and there are **fewer than 250 completed trials**.

#### `CmaEsSampler`
- Used when:
  - The search space is **numerical-only**.
  - There are **more than 250 trials completed**.
  - In the case of more complex numerical spaces, **CMA-ES** is preferred for its evolutionary strategy, and it is **warm-started** with the first 250 completed trials.

### Summary of Rules
- **Multi-Objective**: Always use `NSGAIISampler`.
- **Single-Objective**:
  - **Categorical/Conditional Parameters or Constraints**: Use `TPESampler`.
  - **Numerical Parameters Only**:
    - **< 250 trials**: Use `GPSampler`.
    - **> 250 trials**:
      - If **multiple parameters**: Use `CmaEsSampler` with warm start.
      - Otherwise: Use `TPESampler`.

### Sampler Transition and Initialization
- The process starts with `RandomSampler` to gather preliminary information.
- The sampler transitions based on the optimization state, ensuring that a more suitable sampler is used as trials progress.
- The transition primarily happens in the `before_trial` method, where a suitable sampler is chosen depending on the search space and completed trials.

### Conclusion
The `AutoSampler` ensures that the most appropriate sampler is used by beginning with simpler methods (`RandomSampler`) and gradually switching to more advanced samplers (`TPESampler`, `GPSampler`, `CmaEsSampler`, `NSGAIISampler`) as optimization proceeds. This approach maximizes efficiency and optimizes performance by dynamically adjusting to the optimization problem's requirements.

**Reply by lkaino** [reply]
I'm assuming you replied to my message.

I checked the source code of the AutoSampler and I don't see why it would not use the GPSampler in my case.

I could add some debug prints, but tbh I don't know how does the optuna hub system work. The samplers are not included in the base pip package. Which repos should I clone and install using pip to get access to AutoSampler?

**Reply by not522** [reply]
Hi. Cloning [`optunahub-registry`](https://github.com/optuna/optunahub-registry) and [`load_local_module`](https://optuna.github.io/optunahub/generated/optunahub.load_local_module.html) will help with debugging.

**Reply by y0z** [reply]
The implementation of AutoSampler is available below.
https://github.com/optuna/optunahub-registry/tree/main/package/samplers/auto_sampler

**Comment by SimpleZion** [comment]
Can I customize the parameters of the samplers that might be used within AutoSampler? Some samplers may perform better only after certain parameters are adjusted — for example, enabling the multivariate and group parameters in TPESampler.

**Reply by alan-barzilay** [reply]
hi @SimpleZion! I noticed that [AutoSampler always runs TPESampler on multivariate mode](https://github.com/optuna/optunahub-registry/blob/330f27d9acefc01daceb5a9606c19a39627e843e/package/samplers/auto_sampler/_sampler.py#L140) (not sure if you are still interested in this, sorry for the necromancy haha)

**Comment by alan-barzilay** [comment]
Hi @y0z! I have a few questions about the autosampler and TPE.

0) First, here’s my understanding of how the autosampler works, is this correct?
```

while budget >0:
	if multi-objective:
	    if conditional params → TPE
	    elif few trials completed: # few trials: <250
	        if ≥4 objectives or categorical → TPE
	        else → GP
	    else:
	        if <4 objectives → NSGA-II
	        else → NSGA-III
	else:
	    if categorical or conditional → TPE
	    elif few trials completed → GP
	    elif no constraints → CMA-ES
	    else → TPE 
	    
	budget-=1
```
If this is correct,  I think the documentation should be updated to clarify that the autosampler does not pick a single sampler for the entirety of the study, it swaps between different samplers depending on the number of trials completed so far.

1) I’m curious why CMA-ES is chosen over TPE in that situation. If there’s any literature or discussion that could help me understand this choice, I’d love to read it.

2) More broadly, are there any materials or discussions on how these sampler decisions were made? The sampler comparison table is great, but a bit more documentation on the reasoning behind the choices would be really valuable.

3) Lastly, a small note, I noticed all TPEs have multivariate=True, [which seems to have been the better-performing option for a few years now](https://medium.com/optuna/multivariate-tpe-makes-optuna-even-more-powerful-63c4bfbaebe2). Is there a reason this is not the default yet?

Thank you!

**Reply by y0z** [reply]
Thank you for your comment.

A0: It looks correct at the moment (however, the internal sampler selection rule may change in the future). We will also try to improve the quality of the documentation.

A1: We assume that each sampler has the following properties.
- CMA-ES:
  - Pros: Suitable for numerical optimization. Scalable to the number of evaluations.
  - Cons: Cannot handle non-numerical variables and constraints.
- GP:
  - Pros: Suitable for numerical optimization.
  - Cons: Not suitable for non-numerical variables. Computationally expensive.
- TPE:
  - Pros: Good at non-numerical variables.
  - Cons: In pure numerical optimization, optimization performance seems to be inferior to CMA-ES and GP in many cases.

Then, we use CMA-ES in unconstrained numerical problems when the number of evaluations is large.

A2: We've published [a new article](https://medium.com/optuna/autosampler-full-support-for-multi-objective-constrained-optimization-c1c4fc957ba2), which includes some explanation about the design of the sampler selection rules. We design rules primarily based on the sampler's characteristics and also conduct some internal benchmarking to verify our assumptions. However, good performance on particular benchmarks is relatively unimportant, as we do not want to overfit the benchmark problems we use.

A3. It wasn't easy to change the default sampler or its options, mainly due to compatibility concerns, but we're preparing for that in [the Optuna v5 roadmap](https://medium.com/optuna/optuna-v5-roadmap-ac7d6935a878).
> We aim significantly to enhance the Optuna experience for many users through improvements to the default sampler ...

**Reply by alan-barzilay** [reply]
thank you for the reply @y0z!

I have a quick follow-up question.

> Cons: In pure numerical optimization, optimization performance seems to be inferior to CMA-ES and GP in many cases.

Could you share some cases where TPE underperforms in comparison to GP and CMA-ES?

Also, how do you perform the internal benchmarking? are these tests public somewhere? I would love to look at the work you have all done
